In [1]:
import pymysql
pymysql.install_as_MySQLdb()
from sqlalchemy import create_engine
from sqlalchemy_utils import create_database, database_exists
# Create connection string using credentials following this format
# connection = "dialect+driver://username:password@host:port/database"
connection_str = "mysql+pymysql://root:Nina5056$@localhost/sports_db"

In [2]:
engine = create_engine(connection_str)

In [3]:
# create_database(connection_str)

In [4]:
# Check if the database exists. If not, create it.
if database_exists(connection_str) == False:
  create_database(connection_str)
else:
  print('The database already exists')

The database already exists


In [15]:
import pandas as pd
df = pd.read_excel('Data/customers.xlsx')
df

customer_id         Full Name                     email         State
0          102278       Alica Reary  areary0@sciencedaily.com       Florida
1          102279      Delmor Rubin       drubin1@yahoo.co.jp       Indiana
2          102280       Joanie Hoyt      jhoyt2@bloglovin.com  Pennsylvania
3          102281     Madelena Boat   mboat3@surveymonkey.com        Nevada
4          102282     Sayers Patkin        spatkin4@sogou.com      New York
...           ...               ...                       ...           ...
2842       105120     Letti Ruthven  lruthvenrn@cafepress.com     Tennessee
2843       105121     Averell Plaid         aplaidro@dmoz.org       Florida
2844       105122     Naomi Reynard        nreynardrp@who.int       Arizona
2845       105123  Genvieve Quarrie      gquarrierq@prweb.com      Virginia
2846       105124  Humfrey Whitwood   hwhitwoodrr@archive.org    Washington

[2847 rows x 4 columns]

In [16]:
df.shape

(2847, 4)

In [18]:
df.to_sql('customers', engine, if_exists = 'replace')

2847

In [19]:
import pandas as pd
df2 = pd.read_excel('Data/orders.xlsx')
df2

Date  Order_id  Customer_id       Sport  Revenue  Profit  \
0    2022-01-01     10001       102278    Baseball   183.60   97.29   
1    2022-01-01     10001       102279  Basketball   185.76  103.40   
2    2022-01-01     10001       102280  Basketball   128.16   66.27   
3    2022-01-01     10001       102281      Hockey    45.62   15.46   
4    2022-01-01     10001       102282    Football   106.30   21.75   
...         ...       ...          ...         ...      ...     ...   
2842 2022-12-30     10001       105120    Baseball    10.66    2.04   
2843 2022-12-30     10001       105121    Football    90.20   52.02   
2844 2022-12-31     10001       105122    Football    62.73   37.74   
2845 2022-12-31     10001       105123      Soccer   151.70  114.24   
2846 2022-12-31     10001       105124    Football   139.81  113.22   

      Shipping_cost  Rating  
0                 0     NaN  
1                 0     NaN  
2                 0     NaN  
3                 0     NaN  
4                 0     NaN  
...             ...     ...  
2842              0     4.0  
2843              0     2.0  
2844              0     5.0  
2845              0     NaN  
2846              0     NaN  

[2847 rows x 8 columns]

In [20]:
df2.to_sql('orders', engine, if_exists= 'replace')

2847

In [21]:
q = """SELECT * FROM customers LIMIT 20;"""
pd.read_sql(q, engine)

index  customer_id         Full Name                         email  \
0       0       102278       Alica Reary      areary0@sciencedaily.com   
1       1       102279      Delmor Rubin           drubin1@yahoo.co.jp   
2       2       102280       Joanie Hoyt          jhoyt2@bloglovin.com   
3       3       102281     Madelena Boat       mboat3@surveymonkey.com   
4       4       102282     Sayers Patkin            spatkin4@sogou.com   
5       5       102283      Merwyn Stout            mstout5@sfgate.com   
6       6       102284        Prisca Eve           peve6@wordpress.com   
7       7       102285    Osbourn Bustin           obustin7@oakley.com   
8       8       102286      Abby Sprasen     asprasen8@bizjournals.com   
9       9       102287    Verene Addekin     vaddekin9@theatlantic.com   
10     10       102288     Rainer Norley            rnorleya@slate.com   
11     11       102289     Greta Massard           gmassardb@naver.com   
12     12       102290      Pam Milleton  pmilletonc@independent.co.uk   
13     13       102291    Russell Hyndes         rhyndesd@linkedin.com   
14     14       102292      Tedra Cressy          tcressye@twitter.com   
15     15       102293    Celestina Claw             cclawf@reddit.com   
16     16       102294  Tiffani Taillard       ttaillardg@buzzfeed.com   
17     17       102295    Conroy Coneley  cconeleyh@chicagotribune.com   
18     18       102296      Ivett Clears    iclearsi@elegantthemes.com   
19     19       102297        Jorey Dore          jdorej@posterous.com   

             State  
0          Florida  
1          Indiana  
2     Pennsylvania  
3           Nevada  
4         New York  
5         Michigan  
6            Idaho  
7   South Carolina  
8            Texas  
9            Texas  
10        Virginia  
11         Alabama  
12           Texas  
13      California  
14           Texas  
15      California  
16           Texas  
17      California  
18        New York  
19        New York

In [23]:
q = """SHOW TABLES;"""
pd.read_sql(q, engine)

Tables_in_sports_db
0           customers
1              orders

In [22]:
q = """SELECT * FROM orders LIMIT 20;"""
pd.read_sql(q, engine)

index       Date  Order_id  Customer_id       Sport  Revenue  Profit  \
0       0 2022-01-01     10001       102278    Baseball   183.60   97.29   
1       1 2022-01-01     10001       102279  Basketball   185.76  103.40   
2       2 2022-01-01     10001       102280  Basketball   128.16   66.27   
3       3 2022-01-01     10001       102281      Hockey    45.62   15.46   
4       4 2022-01-01     10001       102282    Football   106.30   21.75   
5       5 2022-01-01     10001       102283    Football    58.11   12.08   
6       6 2022-01-01     10001       102284      Soccer   214.04   70.41   
7       7 2022-01-01     10001       102285  Basketball    37.29    2.55   
8       8 2022-01-01     10001       102286    Baseball    56.19   17.26   
9       9 2022-01-01     10001       102287      Soccer   178.28   56.29   
10     10 2022-01-02     10001       102288    Football    20.94    4.71   
11     11 2022-01-02     10001       102289      Hockey   134.35   47.07   
12     12 2022-01-02     10001       102290    Football   173.17   58.25   
13     13 2022-01-02     10001       102291      Hockey   121.07   33.73   
14     14 2022-01-02     10001       102292  Basketball   210.98   74.53   
15     15 2022-01-02     10001       102293  Basketball   186.97   60.80   
16     16 2022-01-02     10001       102294    Baseball    57.64   10.80   
17     17 2022-01-02     10001       102295      Hockey    14.61    2.82   
18     18 2022-01-02     10001       102296      Soccer    19.39    1.44   
19     19 2022-01-03     10001       102297    Baseball    86.60   16.26   

    Shipping_cost  Rating  
0               0     NaN  
1               0     NaN  
2               0     NaN  
3               0     NaN  
4               0     NaN  
5               0     3.0  
6               0     2.0  
7               0     NaN  
8               0     2.0  
9               0     NaN  
10              0     5.0  
11              0     2.0  
12              0     2.0  
13              0     NaN  
14              0     NaN  
15              0     2.0  
16              0     NaN  
17              0     NaN  
18              0     NaN  
19              0     NaN

# 1.  KPI's:
- Profit
- Total Revenue
- Profit Margin
- Total number of orders

In [28]:
q = """
SELECT 
    SUM(revenue) AS total_revenue,
    SUM(profit) AS total_profit,
    COUNT(*) as total_orders,
    (SUM(profit)/SUM(revenue)) * 100 AS profit_margin
FROM orders;
"""
pd.read_sql(q, engine)

total_revenue  total_profit  total_orders  profit_margin
0      459418.37     284821.86          2847      61.996184

# 2. Now I'd like to query the KPI's above by sport and in decending order of profit margin. 

In [32]:
q = """
SELECT 
    Sport,
    ROUND(SUM(revenue),2) AS total_revenue,
    ROUND(SUM(profit),2) AS total_profit,
    COUNT(*) as total_orders,
    (SUM(profit)/SUM(revenue)) * 100 AS profit_margin
FROM orders
GROUP BY Sport
ORDER BY profit_margin DESC;
"""
pd.read_sql(q, engine)

Sport  total_revenue  total_profit  total_orders  profit_margin
0      Soccer       90157.93      56640.64           561      62.823803
1    Football       94768.32      59328.90           572      62.604149
2    Baseball       95363.99      59698.79           565      62.600978
3  Basketball       92116.23      56275.33           577      61.091656
4      Hockey       87011.90      52878.20           572      60.771228

# 3. How many customers gave a rating?  What was the average rating overall?

- These are rated 1-5.  1 is poor service and 5 being excellent customer service

In [36]:
q = """
SELECT 
    (SELECT COUNT(*) FROM orders WHERE rating is NOT NULL) as number_of_ratings,
    ROUND(AVG(rating),2) AS average_rating
FROM orders;
"""
pd.read_sql(q, engine)

number_of_ratings  average_rating
0               1193            3.13

1193/2847=42% of customers provided a rating.  Average rating was 3.13

# 4.  Order and Group by each rating and show it's revenue, profit, and profit margin.

In [40]:
q = """
SELECT 
    rating,
    SUM(revenue) as total_revenue,
    SUM(profit) as total_profit,
    ROUND((SUM(profit)/SUM(revenue)) * 100, 2) as profit_margin
FROM orders
WHERE rating is not null
GROUP BY rating
ORDER BY rating DESC;
"""
pd.read_sql(q, engine)

rating  total_revenue  total_profit  profit_margin
0     5.0       40565.93      23957.63          59.06
1     4.0       29468.41      17303.72          58.72
2     3.0       38663.38      24209.14          62.62
3     2.0       31838.67      19250.57          60.46
4     1.0       28596.75      16339.88          57.14

- Customers who gave a 3 rating had the highest profit margin. 

# 5.  Find the revenue, profit, and profit margin by State

In [41]:
q = """
SELECT 
    c.state,
    SUM(o.revenue) as total_revenue,
    SUM(o.profit) as total_profit,
    ROUND((SUM(o.profit)/SUM(o.revenue)) * 100, 2) as profit_margin
FROM orders o
join customers c
on o.customer_id = c.customer_id
group by c.state
order by total_profit DESC;
"""
pd.read_sql(q, engine)

state  total_revenue  total_profit  profit_margin
0       California       55470.47      34553.72          62.29
1            Texas       52305.67      32234.74          61.63
2          Florida       36250.92      22398.09          61.79
3         New York       27177.83      17331.45          63.77
4         Virginia       23511.58      14694.96          62.50
5     Pennsylvania       15492.02       9250.66          59.71
6          Georgia       13007.97       7745.32          59.54
7             Ohio       13328.73       7713.25          57.87
8       Washington       11633.95       7262.98          62.43
9          Indiana       11490.83       7012.76          61.03
10        Colorado       11084.28       6717.10          60.60
11       Minnesota       10018.47       6257.23          62.46
12        Missouri        9656.49       6101.95          63.19
13       Louisiana        9696.57       6079.70          62.70
14         Arizona       10186.02       6027.40          59.17
15   Massachusetts        8664.92       6023.10          69.51
16        Illinois        8567.88       5624.77          65.65
17  North Carolina        9388.44       5599.22          59.64
18       Tennessee        8435.80       5363.80          63.58
19         Alabama        8543.43       5358.20          62.72
20     Connecticut        7309.24       4699.25          64.29
21        Kentucky        6972.71       4597.70          65.94
22          Kansas        7671.81       4552.95          59.35
23        Maryland        8044.42       4501.48          55.96
24        Michigan        7163.52       4414.83          61.63
25        Oklahoma        7248.63       4403.45          60.75
26            Utah        5256.80       3656.88          69.56
27            Iowa        5030.35       3367.87          66.95
28      New Jersey        4704.42       2951.39          62.74
29          Nevada        5001.41       2905.47          58.09
30   West Virginia        4354.99       2699.47          61.99
31        Nebraska        3955.96       2607.93          65.92
32       Wisconsin        3972.44       2371.87          59.71
33  South Carolina        4153.66       2355.70          56.71
34      New Mexico        2996.67       2044.45          68.22
35          Oregon        3251.76       2009.55          61.80
36          Alaska        2646.35       1707.98          64.54
37        Delaware        2446.65       1658.72          67.80
38           Idaho        2758.78       1561.04          56.58
39          Hawaii        2116.00       1310.03          61.91
40        Arkansas        2226.27       1256.93          56.46
41   New Hampshire        1497.36       1012.14          67.59
42     Mississippi        1354.42        825.26          60.93
43    South Dakota        1045.06        678.61          64.94
44         Montana         969.45        551.57          56.90
45    North Dakota         705.99        453.54          64.24
46    Rhode Island         559.88        298.68          53.35
47           Maine          91.10         16.72          18.35

- California had the highest profits with Main having the lowest profit.

# 6.  Rank the total revenue, total profit, and profit margin

In [52]:
q = """
SELECT 
    c.state,
    row_number() over (order by SUM(o.revenue) DESC) as revenue_rank,
    SUM(o.revenue) as total_revenue,
    row_number() over (order by SUM(o.profit) DESC) as profit_rank,
    SUM(o.profit) as total_profit,
    row_number() over (order by SUM(o.profit)/SUM(o.revenue) * 100 DESC) as margin_rank,
    ROUND((SUM(o.profit)/SUM(o.revenue)) * 100, 2) as profit_margin
FROM orders o
JOIN customers c
ON o.customer_id = c.customer_id
GROUP BY c.state
ORDER BY revenue_rank;
"""
pd.read_sql(q, engine)

state  revenue_rank  total_revenue  profit_rank  total_profit  \
0       California             1       55470.47            1      34553.72   
1            Texas             2       52305.67            2      32234.74   
2          Florida             3       36250.92            3      22398.09   
3         New York             4       27177.83            4      17331.45   
4         Virginia             5       23511.58            5      14694.96   
5     Pennsylvania             6       15492.02            6       9250.66   
6             Ohio             7       13328.73            8       7713.25   
7          Georgia             8       13007.97            7       7745.32   
8       Washington             9       11633.95            9       7262.98   
9          Indiana            10       11490.83           10       7012.76   
10        Colorado            11       11084.28           11       6717.10   
11         Arizona            12       10186.02           15       6027.40   
12       Minnesota            13       10018.47           12       6257.23   
13       Louisiana            14        9696.57           14       6079.70   
14        Missouri            15        9656.49           13       6101.95   
15  North Carolina            16        9388.44           18       5599.22   
16   Massachusetts            17        8664.92           16       6023.10   
17        Illinois            18        8567.88           17       5624.77   
18         Alabama            19        8543.43           20       5358.20   
19       Tennessee            20        8435.80           19       5363.80   
20        Maryland            21        8044.42           24       4501.48   
21          Kansas            22        7671.81           23       4552.95   
22     Connecticut            23        7309.24           21       4699.25   
23        Oklahoma            24        7248.63           26       4403.45   
24        Michigan            25        7163.52           25       4414.83   
25        Kentucky            26        6972.71           22       4597.70   
26            Utah            27        5256.80           27       3656.88   
27            Iowa            28        5030.35           28       3367.87   
28          Nevada            29        5001.41           30       2905.47   
29      New Jersey            30        4704.42           29       2951.39   
30   West Virginia            31        4354.99           31       2699.47   
31  South Carolina            32        4153.66           34       2355.70   
32       Wisconsin            33        3972.44           33       2371.87   
33        Nebraska            34        3955.96           32       2607.93   
34          Oregon            35        3251.76           36       2009.55   
35      New Mexico            36        2996.67           35       2044.45   
36           Idaho            37        2758.78           39       1561.04   
37          Alaska            38        2646.35           37       1707.98   
38        Delaware            39        2446.65           38       1658.72   
39        Arkansas            40        2226.27           41       1256.93   
40          Hawaii            41        2116.00           40       1310.03   
41   New Hampshire            42        1497.36           42       1012.14   
42     Mississippi            43        1354.42           43        825.26   
43    South Dakota            44        1045.06           44        678.61   
44         Montana            45         969.45           45        551.57   
45    North Dakota            46         705.99           46        453.54   
46    Rhode Island            47         559.88           47        298.68   
47           Maine            48          91.10           48         16.72   

    margin_rank  profit_margin  
0            23          62.29  
1            29          61.63  
2            27          61.79  
3            14          63.77  
4            20          

- Maine had the lowest margin rank and Utah had the highest margin rank
- Maine had the lowest profit rank and CA had the highest profit rank
- Maine had the lowest revenue rank and CA had the highest revenue rank

# 7. Monthly Profits

In [57]:
q = """
WITH monthly_profit AS (
    SELECT 
        MONTH(date) as month,
        SUM(profit) as total_profit
    FROM 
        orders
    GROUP BY MONTH(date)
)
SELECT 
    month,
    total_profit,
    LAG(total_profit) OVER (ORDER BY month) as previous_month_profit,
    ROUND(total_profit - LAG(total_profit) OVER (ORDER BY month),2) as profit_difference
FROM monthly_profit
ORDER BY month;
"""
pd.read_sql(q, engine)

month  total_profit  previous_month_profit  profit_difference
0       1      14013.52                    NaN                NaN
1       2      11244.50               14013.52           -2769.02
2       3      18335.76               11244.50            7091.26
3       4      41131.00               18335.76           22795.24
4       5      38847.24               41131.00           -2283.76
5       6      42802.26               38847.24            3955.02
6       7      31550.40               42802.26          -11251.86
7       8      28681.36               31550.40           -2869.04
8       9      17992.00               28681.36          -10689.36
9      10      13895.44               17992.00           -4096.56
10     11       9760.52               13895.44           -4134.92
11     12      16567.86                9760.52            6807.34